src/docker.c

```diff
@@ -0,0 +1,1754 @@
+        // through and log the error.
+        ERROR("docker: Invalid CPU values for core %d. Used percentage: %lf, "
+              "Proportional System seconds: %lu, CPU seconds: %lu,"
+              "Old seconds: %lu, Num active cores: %d.", i, used_percent,
+              proportional_footprint, delta_percpu,
+              old_stats->old_percpu_usage[i], active_cores);
+      } else {
+        stats[i]->percent_used = used_percent;
+        stats[i]->percent_idle = 100.00 - used_percent;
+      }
+    }
+  } else {
+    unsigned long proportional_footprint = system_cpu_usage/active_cores;
+    for (int i = 0; i < total_cores; i++) {
+      stats[i]->idle = proportional_footprint - stats[i]->usage;
+    }
+    cpu_state_t *old_stats = (cpu_state_t *) calloc(1, sizeof(cpu_state_t));
+    if (old_stats == NULL) {
+      ERROR("docker: compute_cpu_stats. malloc failed! Could not allocate"
+            " historical stats struct for container %s.", container_id);
+      return DOCKER_ERROR;
+    }
+    old_stats->t = now;
+    old_stats->num_cpus = total_cores;
+    old_stats->old_system_usage = system_cpu_usage;
+    old_stats->old_percpu_usage =
+        (unsigned long *) calloc(total_cores, sizeof(unsigned long));
+    if (old_stats->old_percpu_usage == NULL) {
+      ERROR("docker: compute_cpu_stats. malloc failed! Could not allocate"
+            " historical stats list for container %s", container_id);
+      sfree(old_stats);
+      old_stats = NULL;
+      return DOCKER_ERROR;
+    }
+    for (int i = 0; i < total_cores; i++) {
+      old_stats->old_percpu_usage[i] = stats[i]->usage;
+      stats[i]->percent_used = system_cpu_usage > 0 ?
+          (100.00*stats[i]->usage)/proportional_footprint : 0.00;
+      if (stats[i]->percent_used > 100) {
+        ERROR("docker: Invalid CPU values for core %d. Used percentage: %lf, "
+              "Proportional System seconds: %lu, CPU seconds: %lu,"
+              "Num active cores: %d.", i,
+              stats[i]->percent_used, proportional_footprint, stats[i]->usage,
+              active_cores);
+        stats[i]->percent_used = 0.00;
+        stats[i]->percent_idle = 100.00;
+      } else {
+        stats[i]->percent_idle = 100.00 - stats[i]->percent_used;
+      }
+    }
+    if (c_avl_insert(cpu_hist_values, (void *) container_id,
+           (void *) old_stats) < 0) {
+      ERROR("docker: c_avl_insert failed!");
+      return DOCKER_ERROR;
+    }
+  }
+  return 0;
+}
+
+/* Purpose: Computes memory utilization percentage for used and free states by
+ *           the container since the last time the Docker API was queried.
+ *
+ * Params: (memory_stats_t *) stats: memory_stats_t struct which stores memory
+ *                                     metrics.
+ * Returns: int
+ *          0 in case of success.
+ *          -1 in case of error.
+ **/
+static int compute_memory_stats(memory_stats_t *stats) {
+  if (stats == NULL) {
+    ERROR("docker: compute_memory_stats. memory stats NULL");
```

Comment:
Would you want to return an error here? Not 100% how null pointers work in c if you were to proceed.